java/com/google/cose/CoseMessage.java

```diff
@@ -0,0 +1,34 @@
+package com.google.cose;
+
+import co.nstant.in.cbor.model.DataItem;
+import co.nstant.in.cbor.model.Map;
+import com.google.cose.utils.CborUtils;
+
+/**
+ * Implements the base class for COSE Message structure to be implemented for other message types.
+ */
+public abstract class CoseMessage {
+  private final byte[] protectedHeaderBytes;
```

Comment:
Still reading further, but I think it would generally be easier from a use perspective for the library caller to provide a Map and let this library handle serializing it under the hood. I'm generally in favor of passing structured objects through APIs when possible, as opposed to bytes. E.g. I foresee users either bstr encoding the protectedHeaderBytes preemptively when they shouldn't, or vice versa.

Reply:
Actually after further discussion with Paul, just providing two methods for adding protectedHeaders, either a byte array or a Map, should suffice. The latter would immediately canonically serialize the Map into the protectedHeaderBytes field.

Reply:
My only concern was that providing both the methods may cause someone to use both the methods and providing incompatible definitions. Although we can test for that, byte array would make sure that user is providing the actual values they want. But I see your point on preemptive conversion as well.

Reply:
Added protectedHeader map support